kv-cache : separate recurrent vs non-recurrent impl

ggerganov opened the pull request:
### Overview

Attempting to make two separate classes for the 2 types of KV cache:

- `llama_kv_cache_unified   : llama_kv_cache`
- `llama_kv_cache_recurrent : llama_kv_cache`

```mermaid
graph TD;
llama_memory_i --> llama_kv_cache
llama_kv_cache --> llama_kv_cache_unified
llama_kv_cache --> llama_kv_cache_recurrent
```

The main goal of this change is to simplify the logic in the primary `llama_kv_cache_unified` class so that we can more easily extend it with new features such as SWA. Also to introduce a certain level of abstraction that would allow to add new types of KV cache implementations in the future.

### Main changes

- The `llama_context` now operates with the abstract `llama_memory_i` interface.
- Add `llama_memory_params` and use it to implement `llama_model::create_memory()` for creating model-specific cache
- `llama_kv_cache_recurrent` is currently mostly a copy of `llama_kv_cache_unified`, but should be now completely separated and a new recurrent-specific implementation can be done
- Move KV cache shift and defrag code from `llama_context` to `llama_kv_cache_unified`
- The `llama_sbatch` -> `llama_ubatch` logic inside `llama_context:decode()` is now implemented by:
  - `llama_kv_cache::sbatch_init()`
  - `llama_kv_cache::ubatch_next()`

  The thinking is that certain KV cache implementation could require different types of micro-batching (e.g. same-sequence-length ubatch, single-sequence ubatch, etc.)
- Remove `llama_context::output_reorder()` - seemed to be relevant only for recurrent caches. We now have inlined the logic in `llama_context:decode()`
- Remove `llama_context::sbatch`. Instead, create a new one for each decode

### TODO before merge

- [x] Clean-up `llama_kv_cache` interface
- [x] Make `llama_kv_cache_xxx` more private
- [x] Add comments

### Next PRs

- [x] Support cache-less context (needed for embedding-only models such as BERT) (#13108)
- [x] Remove `llama_context_params.logits_all` logic - unnecessary complication, can be achieved with explicit request for logits for all tokens
- [x] Remove `infill` example - obsolete
- [ ] Add proper SWA support to `llama_kv_cache_unified`

### Resolve

- [x] https://github.com/ggml-org/llama.cpp/pull/13113

ggerganov committed:
kv-cache : serparate recurrent vs non-recurrent impl (wip)

ggml-ci

ggerganov committed:
kv-cache : init -> contructor + add llama_memory_params

ggml-ci

ggerganov committed:
kv-cache : fix callback reference

ggml-ci

ggerganov committed:
context : llama_kv_cache -> llama_memory_i

ggml-ci

ggerganov committed:
context : move memory creation logic to model

ggml-ci

ggerganov committed:
llama : remove reference of memory during encode

ggml-ci

ggerganov committed:
kv-cache : hide padding details in the implementation

ggml-ci

ggerganov committed:
kv-cache : add ubatch_next()

ggml-ci

ggerganov committed:
context : simplify sbatch logic

ggml-ci

ggerganov committed:
kv-cache : hide defrag logic in the implementation

ggml-ci

ggerganov committed:
context : hide kv cache details in implementation

ggml-ci

ggerganov committed:
build : fix

ggml-ci

ggerganov committed:
cont : another fix

ggml-ci

ggerganov committed:
kv-cache : simplify interface (wip)

ggml-ci

ggerganov committed:
kv-cache : use separate KV cell structs for unified/recurrent

ggml-ci

ggerganov committed:
kv-cache : clean-up

ggml-ci

ggerganov committed:
model : better llama_model::create_model() signature

ggml-ci

ggerganov committed:
kv-cache : fix recurrent seq_rm()

ggml-ci

slaren commented:
What the reasoning for using `llama_kv_cache` as the base class for `llama_kv_cache_recurrent`? Is there enough code shared between these types to justify this? It seems that there is a lot of complexity in `llama_kv_cache_recurrent` and it would be good if that could be simplified a bit.

On a more general note, I think it is not very usual the way `std::function` callbacks are mixed with inheritance. I think the more typical way to do this would be to create virtual functions that can be overriden in a child class. I wonder if I am missing something here that would prevent implementing this in this way.

ggerganov commented:
> What the reasoning for using llama_kv_cache as the base class for llama_kv_cache_recurrent? Is there enough code shared between these types to justify this? It seems that there is a lot of complexity in llama_kv_cache_recurrent and it would be good if that could be simplified a bit.

The public API currently works with `struct llama_kv_cache *`, so both the recurrent and non-recurrent implementation have to implement it for now.

I think what we need to do in a follow-up PR is:
- Deprecate the public API `llama_kv_cache_`
- Add `llama_memory_` API that works with `struct llama_memory`

At this point, a completely new recurrent-specific implementation can be added: `class llama_memory_recurrent : public llama_memory_i` that would replace the current `llama_kv_cache_recurrent`.

The existing recurrent cache implementation has to be rewritten from scratch, because is was hacked on top of the KV cache implementation by repurposing the K and V tensors for the state space requirements.

---

> On a more general note, I think it is not very usual the way std::function callbacks are mixed with inheritance. I think the more typical way to do this would be to create virtual functions that can be overridden in a child class. I wonder if I am missing something here that would prevent implementing this in this way.

I'll try to update this. Just to make sure, you mean the current:

- `struct llama_kv_cache::callbacks`
- `struct llama_kv_cache::graph_params`

to become interfaces with different implementations based on the type of memory?

slaren commented:
I don't fully understand the code, but I think `get_rope_factors` and `get_buft` could be virtual/abstract functions of `llama_kv_cache`, and a new class can be created if these need a different implementations. But I am not sure that's the case, maybe they can just be regular functions and `llama_kv_cache` should just have a reference to the `llama_model`. `graph_params` looks like it should be an interface, but it feels out of place.

compilade commented:
> The public API currently works with `struct llama_kv_cache *`, so both the recurrent and non-recurrent implementation have to implement it for now.

There will need to be some top-level type which can contain multiple types of KV caches to ease supporting hybrid models. A shared interface for recurrent and non-recurrent state caches is useful to get to that point, at least for maintainability.

The hardest part will be handling errors and properly keeping coherency between the different types of caches (because they don't necessarily roll-back states in the same way). That is relevant mostly for hybrid models, though.

> The existing recurrent cache implementation has to be rewritten from scratch, because is was hacked on top of the KV cache implementation by repurposing the K and V tensors for the state space requirements.

Yes it will need to be rewritten at least to be able to support proper state rollback.

But even if it was repurposing the K and V tensors, there is still some things which I think will remain, since Mamba and RWKV *do* have 2 types of recurrent states per layer.

ggerganov committed:
kv-cache : replace `struct callbacks` with `llama_model &`

ggml-ci

ggerganov committed:
kv-cache : replace `struct graph_params` with `llama_context &`

ggml-ci

ggerganov commented:
@slaren In [7e4b545](https://github.com/ggml-org/llama.cpp/pull/12799/commits/7e4b54596dd846f976b169120842eeba5c416f82) I replaced the `struct callbacks` by maintaining a reference of `llama_model` in the `llama_kv_cache` implementation. And in [73df685](https://github.com/ggml-org/llama.cpp/pull/12799/commits/73df685426b93cdf279336a8a016bb4197eb8981) I replaced the `struct graph_params` by passing a reference to the `llama_context`.

PTAL if you think these changes are good.

slaren reviewed (commented):
The changes look good. While testing this, I noticed that the KV cache is always allocated on the CPU.

ggerganov committed:
kv-cache : fix offload check

ggml-ci

ggerganov committed:
context : avoid passing unique_ptr

ggml-ci

ggerganov commented on src/llama-context.h:
```diff
@@ -27,7 +27,15 @@ struct llama_context {
+    ggml_backend_sched_t get_sched() const;
+
+    ggml_context * get_ctx_compute() const;
+
+    // TODO: this method might be possible to avoid (seach for TAG_BACKENDS)
+    const std::vector<ggml_backend_ptr> & get_backends() const;
```
I was wondering if we can avoid exposing the backends here. Currently, this is only needed in the `llama_kv_cache_unified::build_rope_shift()` method to determine the backend of the dequantized cache tensors:


https://github.com/ggml-org/llama.cpp/blob/ed7bb58a9eb811ccbcc7c659aa7ec5bbee9109e7/src/llama-kv-cache.cpp#L626-L637

@slaren Any ideas how to make this simpler?

ggerganov replied:
Thanks, fixed now.

ggerganov committed:
kv-cache : avoid using the backends from the llama_context

ref #13113

ggml-ci

ggerganov committed:
kv-cache : more consistent debug logs [no ci]

ggerganov committed:
kv-cache : do not pass the full llama_context for kv graphs

ggml-ci

slaren commented on src/llama-kv-cache.cpp:
```diff
@@ -765,70 +591,333 @@ size_t llama_kv_cache_unified::size_v_bytes() const {
-    //  if ids[i] == i || ids[i] == n_kv, then cell i is not moved
-    //
-    auto & ids = defrag_info.ids;
+        tmp = ggml_rope_ext_inplace(ctx, tmp,
```
Not necessary to use an inplace operation here (from #13113).
```suggestion
        tmp = ggml_rope_ext(ctx, tmp,
```

slaren commented on src/llama-kv-cache.cpp:
```diff
@@ -1174,7 +1258,1034 @@ bool llama_kv_cache_unified::state_read_data(llama_io_read_i & io, uint32_t cell
         return false;
     }
-    if (v_trans != (bool) v_trans) {
+    if (this->v_trans != (bool) v_trans) {
```
The diff is not very easy to follow here, is `state_read_data` now duplicated in `llama_kv_cache_unified` and in `llama_kv_cache_recurrent`?

slaren commented on src/llama-model.cpp:
```diff
@@ -12775,26 +12788,29 @@ llama_memory_i * llama_model::create_memory() const {
-                });
+                res = new llama_kv_cache_recurrent(
+                        *this,
+                        GGML_TYPE_F32,
+                        GGML_TYPE_F32,
```
Is it intentional to always use F32 here?

compilade replied:
Yes, the Mamba and RWKV operators currently expect the states to be in F32.

compilade commented on src/llama-kv-cache.cpp:
```diff
@@ -1174,7 +1258,1034 @@ bool llama_kv_cache_unified::state_read_data(llama_io_read_i & io, uint32_t cell
+
+    //////////////////////////////////////////////
+    // TODO: this should not mutate the KV cache !
+    kv_cell & cell = const_cast<kv_cell &>(cells[i]);
```
```suggestion
    kv_cell & cell = const_cast<kv_cell &>(cells[cell_id]);
```

Otherwise multi-user inference is broken for recurrent models. See <https://github.com/ggml-org/llama.cpp/pull/9126#issuecomment-2846166195>.

compilade commented on src/llama-kv-cache.cpp:
```diff
@@ -1174,7 +1258,1034 @@ bool llama_kv_cache_unified::state_read_data(llama_io_read_i & io, uint32_t cell
+
+    //////////////////////////////////////////////
+    // TODO: this should not mutate the KV cache !
+    kv_cell & cell = const_cast<kv_cell &>(cells[i]);
```
```suggestion
    kv_cell & cell = const_cast<kv_cell &>(cells[cell_id]);
```
Same, this should fix multi-user inference.

ggerganov replied:
We should add a small multi-user test with a recurrent model to `server/tests` to be able to spot such regressions.

ggerganov committed:
kv-cache : remove comment

ggerganov committed:
kv-cache : ggml_rope_ext_inplace -> ggml_rope_ext

ggml-ci

ggerganov committed:
kv-cache : fix recurrent multi-user case

ggml-ci

ggerganov commented:
@slaren @compilade I think this should be good to merge - any additional comments?

ggerganov committed:
memory : remove comments [no ci]

ggerganov commented:
> There will need to be some top-level type which can contain multiple types of KV caches to ease supporting hybrid models. A shared interface for recurrent and non-recurrent state caches is useful to get to that point, at least for maintainability.
> 
> The hardest part will be handling errors and properly keeping coherency between the different types of caches (because they don't necessarily roll-back states in the same way). That is relevant mostly for hybrid models, though.

I think that when we introduce the `llama_memory_` API (see https://github.com/ggml-org/llama.cpp/pull/12799#issuecomment-2839747983) we can redesign how the caches are used. The existing `llama_kv_cache_seq_` API is not great in general (error prone and a bit hacky to use), so it would be a good opportunity to think about ways to simplify and improve it.

compilade commented on src/llama-context.cpp:
```diff
@@ -1424,18 +1054,52 @@ int llama_context::decode(llama_batch & inp_batch) {
 
-        if (sorted_output) {
-            sbatch.out_ids.clear();
+        // make the outputs have the same order they had in the user-provided batch
+        // note: this is mostly relevant for recurrent models atm
```
It's also *only* relevant when using `get_embeddings`, because the buffer in that case has to be ordered to keep the API backward compatible. When purely using `get_embeddings_ith`, it's not required.

Unconditionally sorting is unnecessary and is likely slower. Also it seems like some assertions here break multi-user inference for recurrent models (since the line right after this block where `n_outputs = n_outputs_all` is assumed to have run before the sorting routine, but it hasn't).

ggerganov replied:
The main reason to decide to always reorder is because otherwise we have to maintain the `sbatch` in the state of the context. This introduces some complexity that is hard to reason around so I decided to take the hit.

We should add a test that exercises this branch. What is a server scenario that would trigger the reordering?

I'll PR the `n_outputs = n_outputs_all` before the sorting fix.
